pkgs/servers/traefik/default.nix

```
@@ -1,39 +1,37 @@
-{ stdenv, buildGoPackage, fetchFromGitHub, bash, go-bindata}:
+{ stdenv, buildGoModule, fetchFromGitHub, go-bindata }:
 
-buildGoPackage rec {
+buildGoModule rec {
   pname = "traefik";
-  version = "1.7.14";
-
-  goPackagePath = "github.com/containous/traefik";
+  version = "2.1.1";
 
   src = fetchFromGitHub {
     owner = "containous";
     repo = "traefik";
     rev = "v${version}";
-    sha256 = "1j3p09j8rpdkp8v4d4mz224ddakkvhzchvccm9qryrqc2fq4022v";
+    sha256 = "116w1147an7xfhg4x9rbqkhv7c6si1np0cr3s65287wakc2r65pw";
   };
 
-  buildInputs = [ go-bindata bash ];
-
-  buildPhase = ''
-    runHook preBuild
-    (
-      cd go/src/github.com/containous/traefik
-      bash ./script/make.sh generate
-
-      CODENAME=$(awk -F "=" '/CODENAME=/ { print $2}' script/binary)
-      go build -ldflags "\
-        -X github.com/containous/traefik/version.Version=${version} \
-        -X github.com/containous/traefik/version.Codename=$CODENAME \
-      " -a -o $bin/bin/traefik ./cmd/traefik
-    )
-    runHook postBuild
+  goPackagePath = "github.com/containous/traefik";
+
+  modSha256 = "1cqjayxngw16g7yfksqvg4aaanmwqwiqlb1caplcgbgzwf5z4vh1";
+  subPackages = [ "cmd/traefik" ];
+
+  buildInputs = [ go-bindata ];
+
+  preBuild = ''
+    go generate
+
+    CODENAME=$(awk -F "=" '/CODENAME=/ { print $2}' script/binary)
+
+    export buildFlagsArray="-ldflags=\
+      -X github.com/containous/traefik/version.Version=${version} \
+      -X github.com/containous/traefik/version.Codename=$CODENAME"
   '';
```

Please move these out of the `preBuild` phase. Example :point_down: https://github.com/NixOS/nixpkgs/blob/75a9868dbe66bff430d3e1c01b5067a1c48053c3/pkgs/applications/networking/cluster/helm/default.nix#L16

---

I had them there because of the `CODENAME` variable, which gets evaluated at the `preBuild` phase. Are you sure that `buildFlagsArray` is going to be evaluated properly?

---

I'll side with @jokogr: `buildFlagsArray` can't be used here. The best we can do is use `makeFlagsArray`, which is what the nixpkgs manual recommends:

```suggestion
    makeFlagsArray+=("-ldflags=\
      -X github.com/containous/traefik/version.Version=${version} \
      -X github.com/containous/traefik/version.Codename=$CODENAME")

```